File: src/main/java/com/cxxcxx/zinecraft/core/registry/ModBlockEntity.java
package com.cxxcxx.zinecraft.core.registry;

import com.cxxcxx.zinecraft.api.registry.builder.BlockBuilder;
import com.cxxcxx.zinecraft.api.registry.builder.BlockEntityBuilder;
import com.cxxcxx.zinecraft.core.Zinecraft;
import com.cxxcxx.zinecraft.core.entity.ExampleBlockEntity;
import com.cxxcxx.zinecraft.core.entity.ExampleEntityBlock;
import net.minecraft.world.level.block.SoundType;
import net.minecraft.world.level.block.state.BlockBehaviour.Properties;

public final class ModBlockEntity {
  private ModBlockEntity() {
  }

  public static final BlockEntityBuilder<ExampleBlockEntity, ExampleEntityBlock> EXAMPLE_BLOCK_ENTITY =
      new BlockEntityBuilder<>(
          Zinecraft.BLOCK_ENTITIES,
          "example_block_entity",
          ExampleBlockEntity::new,
          new BlockBuilder<>(
              Zinecraft.BLOCKS,
              "example_entity_block",
              "示例实体方块",
              () -> new ExampleEntityBlock(Properties.of().sound(SoundType.GRASS))
          )
      ).build();

  public static void bootstrap() {
  }
}

File: core/src/main/kotlin/com/cxxcxx/zinecraft/core/entity/ModBlockEntity.kt
package com.cxxcxx.zinecraft.core.entity


import com.cxxcxx.zinecraft.core.ZinecraftCore
import com.cxxcxx.zinecraft.core.block.ModBlock
import net.minecraft.core.Registry
import net.minecraft.core.registries.BuiltInRegistries
import net.minecraft.resources.ResourceLocation
import net.minecraft.world.level.block.Block
import net.minecraft.world.level.block.entity.BlockEntity
import net.minecraft.world.level.block.entity.BlockEntityType
import net.minecraft.world.level.block.entity.BlockEntityType.BlockEntitySupplier


object ModBlockEntity {

  val EXAMPLE_BLOCK_ENTITY: BlockEntityType<ExampleBlockEntity> =
    register(
      "example_block_entity", ::ExampleBlockEntity,
      ModBlock.EXAMPLE_ENTITY_BLOCK
    )


  private fun <T : BlockEntity> register(
    name: String,
    entityFactory: BlockEntitySupplier<out T>,
    vararg blocks: Block
  ): BlockEntityType<T> {
    val id = ResourceLocation.fromNamespaceAndPath(ZinecraftCore.MOD_ID, name)
    return Registry.register(
      BuiltInRegistries.BLOCK_ENTITY_TYPE,
      id,
      BlockEntityType.Builder.of<T>(entityFactory, *blocks).build()
    )
  }

  fun init() {
    EXAMPLE_BLOCK_ENTITY
  }
}